### PR TITLE
Support column split in GPU predictor

### PR DESCRIPTION
**Pull request by rongou**
The approach is similar to the CPU predictor: first collect all the decisions at each node and whether the feature is missing locally, then do an allreduce on the bit vectors, and finally use the bit vectors to find the leaf.

Still need to do more testing in a followup.

### CODE REVIEW

**Comment by rongou** [comment]
@trivialfis

File: src/predictor/cpu_predictor.cc [hunk 1]
```diff
@@ -511,7 +504,7 @@ class ColumnSplitHelper {
     if (missing_bits_.Check(bit_index)) {
       return node.DefaultChild();
     } else {
-      return node.LeftChild() + decision_bits_.Check(bit_index);
+      return node.LeftChild() + !decision_bits_.Check(bit_index);
```
Comment (trivialfis) [inline src/predictor/cpu_predictor.cc:507]: Is this a fix, or the meaning of the bits are reversed?

File: tests/cpp/predictor/test_gpu_predictor.cu [hunk 2]
```diff
@@ -57,6 +57,72 @@ TEST(GPUPredictor, Basic) {
   }
 }
 
+namespace {
+void VerifyBasicColumnSplit(std::array<std::vector<float>, 32> const& expected_result) {
+  auto const world_size = collective::GetWorldSize();
+  auto const rank = collective::GetRank();
+
+  auto lparam = MakeCUDACtx(rank);
```
Comment (trivialfis) [inline tests/cpp/predictor/test_gpu_predictor.cu:65]: Let's keep the naming `ctx`, easier for me to grep ;-)

File: tests/cpp/predictor/test_gpu_predictor.cu [hunk 3]
```diff
@@ -57,6 +57,72 @@ TEST(GPUPredictor, Basic) {
...
+  std::unique_ptr<Predictor> predictor =
+      std::unique_ptr<Predictor>(Predictor::Create("gpu_predictor", &lparam));
+  predictor->Configure({});
+
+  for (size_t i = 1; i < 33; i *= 2) {
+    size_t n_row = i, n_col = i;
+    auto dmat = RandomDataGenerator(n_row, n_col, 0).GenerateDMatrix();
+    std::unique_ptr<DMatrix> sliced{dmat->SliceCol(world_size, rank)};
+
+    Context ctx;
```
Comment (trivialfis) [inline tests/cpp/predictor/test_gpu_predictor.cu:75]: `MakeCUDACtx(rank)`.

File: tests/cpp/predictor/test_gpu_predictor.cu [hunk 4]
```diff
@@ -57,6 +57,72 @@ TEST(GPUPredictor, Basic) {
...
+}
+}  // anonymous namespace
+
+TEST(GPUPredictor, MGPUBasicColumnSplit) {
+  auto const n_gpus = common::AllVisibleGPUs();
+  if (n_gpus <= 1) {
+    GTEST_SKIP() << "Skipping MGPUIBasicColumnSplit test with # GPUs = " << n_gpus;
+  }
+
+  auto lparam = MakeCUDACtx(0);
```
Comment (trivialfis) [inline tests/cpp/predictor/test_gpu_predictor.cu:98]: `ctx`.

File: tests/cpp/predictor/test_gpu_predictor.cu [hunk 5]
```diff
@@ -57,6 +57,72 @@ TEST(GPUPredictor, Basic) {
...
+  std::unique_ptr<Predictor> predictor =
+      std::unique_ptr<Predictor>(Predictor::Create("gpu_predictor", &lparam));
+  predictor->Configure({});
+
+  std::array<std::vector<float>, 32> result{};
+  for (size_t i = 1; i < 33; i *= 2) {
+    size_t n_row = i, n_col = i;
+    auto dmat = RandomDataGenerator(n_row, n_col, 0).GenerateDMatrix();
+
+    Context ctx;
```
Comment (trivialfis) [inline tests/cpp/predictor/test_gpu_predictor.cu:108]: `MakeCUDACtx(0)`.

File: tests/cpp/predictor/test_gpu_predictor.cu [hunk 6]
```diff
@@ -57,6 +57,72 @@ TEST(GPUPredictor, Basic) {
...
+    GTEST_SKIP() << "Skipping MGPUIBasicColumnSplit test with # GPUs = " << n_gpus;
+  }
+
+  auto lparam = MakeCUDACtx(0);
+  std::unique_ptr<Predictor> predictor =
+      std::unique_ptr<Predictor>(Predictor::Create("gpu_predictor", &lparam));
+  predictor->Configure({});
+
+  std::array<std::vector<float>, 32> result{};
+  for (size_t i = 1; i < 33; i *= 2) {
```
Comment (trivialfis) [inline tests/cpp/predictor/test_gpu_predictor.cu:104]: What is this iterating over? Are you running it many times to check race conditions?

File: src/predictor/gpu_predictor.cu [hunk 7]
```diff
@@ -720,7 +917,8 @@ class GPUPredictor : public xgboost::Predictor {
   }
 
  public:
-  explicit GPUPredictor(Context const* ctx) : Predictor::Predictor{ctx} {}
+  explicit GPUPredictor(Context const* ctx)
+      : Predictor::Predictor{ctx}, column_split_helper_{ctx->gpu_id} {}
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:921]: Can we pass the `ctx` instead? It contains a `CUDAContext`, which we can query for stream. It doesn't matter at the moment, but we are moving away from the legacy stream, would be great if the new code can use the context by default.

File: src/predictor/predict_fn.h [hunk 8]
```diff
@@ -7,20 +7,26 @@
 #include "xgboost/tree_model.h"
 
 namespace xgboost::predictor {
+/** @brief Whether it should traverse to the left branch of a tree. */
+template <bool has_categorical>
+inline XGBOOST_DEVICE bool GetDecision(RegTree::Node const &node, bst_node_t nid, float fvalue,
```
Comment (trivialfis) [inline src/predictor/predict_fn.h:12]: No need for `inline`. template function is inlined.

File: src/predictor/cpu_predictor.cc (hunk 1 above)
Comment (rongou) [reply src/predictor/cpu_predictor.cc:507]: The meaning is reversed to match `common::Decision()`.

File: src/predictor/gpu_predictor.cu (hunk 7 above)
Comment (rongou) [reply src/predictor/gpu_predictor.cu:921]: Done.

File: src/predictor/predict_fn.h (hunk 8 above)
Comment (rongou) [reply src/predictor/predict_fn.h:12]: Done.

File: tests/cpp/predictor/test_gpu_predictor.cu (hunk 2 above)
Comment (rongou) [reply tests/cpp/predictor/test_gpu_predictor.cu:65]: Done.

File: tests/cpp/predictor/test_gpu_predictor.cu (hunk 3 above)
Comment (rongou) [reply tests/cpp/predictor/test_gpu_predictor.cu:75]: Done.

File: tests/cpp/predictor/test_gpu_predictor.cu (hunk 4 above)
Comment (rongou) [reply tests/cpp/predictor/test_gpu_predictor.cu:98]: Done.

File: tests/cpp/predictor/test_gpu_predictor.cu (hunk 6 above)
Comment (rongou) [reply tests/cpp/predictor/test_gpu_predictor.cu:104]: Yeah I guess. This was just copied from the `Basic` test.

File: tests/cpp/predictor/test_gpu_predictor.cu (hunk 5 above)
Comment (rongou) [reply tests/cpp/predictor/test_gpu_predictor.cu:108]: Done.

File: src/predictor/cpu_predictor.cc (hunk 1 above)
Comment (trivialfis) [reply src/predictor/cpu_predictor.cc:507]: Could you please the comment for `ColumnSplitHelper`?

File: src/predictor/gpu_predictor.cu [hunk 9]
```diff
@@ -622,6 +621,199 @@ size_t SharedMemoryBytes(size_t cols, size_t max_shared_memory_bytes) {
...
+          MaskBitVectorKernel, data, model.nodes.ConstDeviceSpan(),
+          model.tree_segments.ConstDeviceSpan(), model.tree_group.ConstDeviceSpan(),
+          model.split_types.ConstDeviceSpan(), model.categories_tree_segments.ConstDeviceSpan(),
+          model.categories_node_segments.ConstDeviceSpan(), model.categories.ConstDeviceSpan(),
+          decision_bits, missing_bits, model.tree_beg_, model.tree_end_, num_features, num_rows,
+          entry_start, num_nodes, use_shared, nan(""));
+
+      AllReduceBitVectors(&decision_storage, &missing_storage);
+
+      dh::LaunchKernel {grid, kBlockThreads} (
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:779]: Now that you have the context, could you please help add the boilerplate code:
``` c++
#include "../common/cuda_context.cuh"

auto cuctx = ctx->CUDAContext();
d::LaunchKernek{..., cuctx->Stream()}();
```

File: tests/cpp/predictor/test_gpu_predictor.cu (hunk 6 above)
Comment (trivialfis) [reply tests/cpp/predictor/test_gpu_predictor.cu:102]: Time flies. You wrote the basic test in #4749

File: src/predictor/gpu_predictor.cu [hunk 10]
```diff
@@ -622,6 +622,205 @@ size_t SharedMemoryBytes(size_t cols, size_t max_shared_memory_bytes) {
...
+  static void ResizeBitVectors(dh::caching_device_vector<BitType>* decision_storage,
+                               dh::caching_device_vector<BitType>* missing_storage,
+                               std::size_t total_bits) {
+    auto const size = BitVector::ComputeStorageSize(total_bits);
+    auto const old_decision_size = decision_storage->size();
+    if (old_decision_size < size) {
+      decision_storage->resize(size);
+    }
+    if (old_decision_size != 0) {
+      thrust::fill(decision_storage->begin(), decision_storage->end(), 0);
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:811]: Consider using the thrust policy: `ctx_->CUDACtx()-CTP()`

File: src/predictor/gpu_predictor.cu [hunk 11]
```diff
@@ -622,6 +622,205 @@ size_t SharedMemoryBytes(size_t cols, size_t max_shared_memory_bytes) {
...
+    }
+    auto const old_missing_size = missing_storage->size();
+    if (old_missing_size < size) {
+      missing_storage->resize(size);
+    }
+    if (old_missing_size != 0) {
+      thrust::fill(missing_storage->begin(), missing_storage->end(), 0);
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:818]: Consider using the thrust policy: `ctx_->CUDACtx()-CTP()`

File: src/predictor/gpu_predictor.cu [hunk 12]
```diff
@@ -622,6 +622,205 @@ size_t SharedMemoryBytes(size_t cols, size_t max_shared_memory_bytes) {
...
+
+  static void ResizeBitVectors(dh::caching_device_vector<BitType>* decision_storage,
+                               dh::caching_device_vector<BitType>* missing_storage,
+                               std::size_t total_bits) {
+    auto const size = BitVector::ComputeStorageSize(total_bits);
+    auto const old_decision_size = decision_storage->size();
+    if (old_decision_size < size) {
+      decision_storage->resize(size);
+    }
+    if (old_decision_size != 0) {
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:810]: Why do you use `old_decision_size` instead of `size`?

File: src/predictor/cpu_predictor.cc (hunk 1 above)
Comment (trivialfis) [reply src/predictor/cpu_predictor.cc:507]: That makes sense.

**Comment by rongou** [comment]
@trivialfis Can this be merged? Thanks!

File: src/predictor/gpu_predictor.cu (hunk 12 above)
Comment (trivialfis) [reply src/predictor/gpu_predictor.cu:810]: Could you please share some insight on this?

Comment (trivialfis) [reply src/predictor/gpu_predictor.cu:810]: @rongou

**Review by trivialfis — APPROVED** [review]

**Pull request merged**